### PR TITLE
Fix local deps

### PR DESCRIPTION
**Pull request by fkleuver**
# Pull Request

## 📖 Description

This PR fixes locally registered dependencies in jit mode (and a small bit of test cleanup).

- `createRenderContext` is called [much earlier](https://github.com/aurelia/aurelia/compare/fix-local-deps?expand=1#diff-b12a454f7d84dd07c94592e2fb3b0a30R257) and is also passed into the template compiler so that the resources are available throughout. 
- the render context becomes a mandatory argument in a couple places where it was optional before (but it may be `null`), making the parameters a bit more visible. 
- render context is also passed via `hydrate` now so that you actually get a proper container tree representing the resource tree, instead of only sibling containers based on the root like it was before
- A special `resourceLookup` object is added to the container where any resolvers with string keys are stored. This lookup is cloned into child containers so that the resource model in the template compiler doesn't need to traverse up the container tree

Not necessarily an ideal implementation, a bit hacky/leaky from the perspective of `RuntimeCompilationResources`. But I would really have to sit down for this to come up with something better, and I think other things are a bit higher priority than getting this "proper" right now. So we may need to revisit this bit in the future.

### 🎫 Issues

Registering resources via the decorator didn't work

## 👩‍💻 Reviewer Notes

I believe this PR is small enough (for a change.. :) ) that looking at the files would give a pretty good picture. Besides the irrelevant test cleanup, of course.

## 📑 Test Plan

There is one very lonely little test: https://github.com/aurelia/aurelia/compare/fix-local-deps?expand=1#diff-af2b863bb67c21ecaf1b2faf0ad4eb3eR1227
Again, this needs to be more thorough (especially since all kinds of pesky situations can arise from even the smallest flaw in the container) but all else seems to work, so leaving it be for now. Kernel tests need to be improved overall anyway.

## ⏭ Next Steps

With relation to this issue, add a few hundred edge case tests and see where that leads in terms of fixes/improvements. I don't necessarily want to redesign this over theoretical issues, but actually prove that there are issues via tests. I'm hoping @EisenbergEffect can help me with some examples and cases in the future here.

### CODE REVIEW

File: packages/kernel/src/di.ts [hunk 1]
```diff
@@ -508,6 +509,7 @@ export class Factory implements IFactory {
 /** @internal */
 export interface IContainerConfiguration {
   factories?: Map<Function, IFactory>;
+  resourceLookup?: Record<string, IResourceType<unknown, unknown>>;
```
[the remark this answers is not in the document]
Comment (fkleuver) [reply packages/kernel/src/di.ts]: Not required no, but performant. The only alternative I see right now is to let the resource thing traverse up to the parent container. The number of jumps can add up with deeply nested resources, and so can the number of calls with granular components.

I don't feel very strongly about it and am willing to look into other options, but I do want to point something out.
I don't think `IResourceType` is a templating concern per se. I think it is a DI concern if we look at vNext and forget about vCurrent for a minute.
A resource is a very broad concept; in our case, it's nothing but a class that has a `kind` (which is merely a typing for the class and definition) and a `description` (which is at its core an interface that has nothing but a string name).

I think string-keyed resources can indeed be far broader than templating or even Aurelia. They can be files, uri's, guids with entities, really anything. And DI this way offers a performant way of storing/accessing a very large number of them.

That's just my 2c.

[the remark this answers is not in the document]
Comment (EisenbergEffect) [reply packages/kernel/src/di.ts]: True. That's a good argument. Let's get this in and then we may want to bench a few variations on how things are handled internally later on, just to make sure we're still performing well for other DI scenarios.
